MapleServer2/Types/Player.cs

```diff
@@ -233,5 +238,83 @@ public void Warp(MapPlayerSpawn spawn, int mapId)
             Rotation = spawn.Rotation.ToFloat();
             Session.Send(FieldPacket.RequestEnter(Session.FieldPlayer));
         }
+
+        public void SetPrestigeLevel(int level)
+        {
+            PrestigeLevel = level;
+            Session.Send(PrestigePacket.SendPrestigeLevelUp(Session.FieldPlayer, PrestigeLevel));
+        }
+
+        public void SetLevel(short level)
+        {
+            Level = level;
+            Experience = 0;
+            Session.Send(ExperiencePacket.SendExpUp(0, Experience, 0));
+            Session.Send(ExperiencePacket.SendLevelUp(Session.FieldPlayer, Level));
+        }
+
+        public void GainExp(int amount)
+        {
+            if (amount <= 0 || Level >= MAX_LEVEL)
+            {
+                return;
+            }
+
+            long newExp = Experience + amount + RestExp;
+
+            if (RestExp > 0)
+            {
+                RestExp -= amount;
+
+            }
+            else
+            {
+                RestExp = 0;
+            }
+
+            while (newExp >= ExpMetadataStorage.GetExpToLevel(Level))
+            {
+                newExp -= ExpMetadataStorage.GetExpToLevel(Level);
+                HandleLevelUp();
+                if (Level >= MAX_LEVEL)
+                {
+                    newExp = 0;
+                    break;
+                }
+            }
+
+            Experience = newExp;
+            Session.Send(ExperiencePacket.SendExpUp(amount, newExp, RestExp));
+        }
+
+        private void HandleLevelUp()
+        {
+            Level++;
+            // TODO: Gain max HP and heal to max hp
+            StatPointDistribution.AddTotalStatPoints(5);
+            Session.Send(StatPointPacket.WriteTotalStatPoints(this));
+            Session.Send(ExperiencePacket.SendLevelUp(Session.FieldPlayer, Level));
+        }
+
+        public void GainPrestigeExp(long amount)
+        {
+            if (Level < 50) // Can only gain prestige exp after level 50.
+            {
+                return;
+            }
+            // Prestige exp can only be earned 1M exp per day. 
+            // TODO: After 1M exp, reduce the gain and reset the exp gained every midnight.
+
+            long newPrestigeExp = PrestigeExp + amount;
+
+            if (newPrestigeExp >= 1000000)
+            {
+                newPrestigeExp -= 1000000;
+                PrestigeLevel++;
+                Session.Send(PrestigePacket.SendPrestigeLevelUp(Session.FieldPlayer, PrestigeLevel));
+            }
+            PrestigeExp = newPrestigeExp;
+            Session.Send(PrestigePacket.SendPrestigeExpUp(this, amount));
+        }
```

Comment:
This should perhaps be moved to an Experience/Level type class separate from player

Reply:
I was thinking about moving it to PlayerStats but that struct can't be change since its used for a packet structure. You have any idea where or what it should be called?